File: golang/architecture-shell-cqrs/command.go
package cqrs

// Command represents state-changing operations with no return value.
// Commands MUST modify state (otherwise use Query).
// Commands MUST execute within transaction boundary (enforced by UnitOfWork behavior).
// Commands MAY return Result[Unit] for business validation errors.
type Command interface {
	BaseRequest
	// IsCommand is a marker method to identify commands.
	IsCommand()
}

// CommandOf represents state-changing operations with return value.
// Examples: created entity ID, updated version number.
// Commands MUST modify state (otherwise use Query).
// Commands MUST execute within transaction boundary.
// TResult typically Result[TValue] for mixed error handling.
// Return values should be minimal (IDs, counts, versions) not full entities.
type CommandOf[TResult any] interface {
	BaseRequest
	IsCommand()
	// GetResultType returns a zero value of TResult for runtime type checking.
	GetResultType() TResult
}

File: typescript-nodejs/architecture-shell-cqrs/src/Command.ts
import { BaseRequest } from './BaseRequest';

/**
 * Command - Marker interface for state-changing operations with no return value
 *
 * Commands represent application-layer operations that modify state.
 * They return void (Unit equivalent) or Result<void> for business validation errors.
 *
 * @remarks
 * Constraints:
 * - MUST modify state (otherwise use Query)
 * - MUST execute within transaction boundary (enforced by UnitOfWork behavior)
 * - MAY return Result<void> for business validation errors
 * - Infrastructure errors (DB down) should throw exceptions
 *
 * @example
 * ```typescript
 * export class PublishBlogPostCommand implements Command {
 *   _isCommand = true;
 *   constructor(
 *     public readonly blogPostId: string,
 *     public readonly publishDate: Date
 *   ) {}
 * }
 * ```
 *
 * @see {@link Query} for read-only operations
 */
export interface Command extends BaseRequest<void> {
  _isCommand?: true;
}
